### PR TITLE
Add a microbench bench test to detect performance regressions

### PR DESCRIPTION
**Pull request by criccomini**
(Creating as draft to get some feedback)

@dmvk detected a performance regression that's been discussed in #371 and #370. A small criterion test was created to illustrate the issue.

https://github.com/dmvk/slatedb-microbench/blob/0.2/Cargo.toml

In this PR, I'm:

- Adding `criterion` to SlateDB'
- Importing @dmvk's test into a `benches` directory

I ran the test locally on my laptop and got:

```
     Running benches/db_operations.rs (target/release/deps/db_operations-1853152ecc617580)
put                     time:   [9.1050 µs 9.1298 µs 9.1607 µs]
Found 93 outliers among 1000 measurements (9.30%)
  12 (1.20%) low severe
  11 (1.10%) low mild
  35 (3.50%) high mild
  35 (3.50%) high severe
```

_#250 did some work to support Criterion and microbenchmarks, bet we haven't merged it yet. If we merge this, that code will need to be rebased._

### CODE REVIEW

**Comment by criccomini** [comment]
After this PR is merged, I'd like to create a follow-on PR that sets a baseline against main and adds a Github PR Check for pull requests that compares the branch to the baseline and fails if it exceeds some threshold.

**Review by agavra — APPROVED** [review]
This LGTM - ran it locally with `cargo bench` to make sure it works in my setup as well.

Criterion is neat! I miss my days with JMH. Just out of curiosity I tweaked the code to run with various different key/value sizes and the performance of this benchmark is about the same no matter the input:
```
     Running benches/db_operations.rs (target/release/deps/db_operations-d7e809828db4f296)
db_put_non_durable/put/key_16B_value_64B
                        time:   [6.9504 µs 7.0479 µs 7.1605 µs]
                        thrpt:  [139.66 Kelem/s 141.89 Kelem/s 143.88 Kelem/s]
Found 108 outliers among 1000 measurements (10.80%)
  41 (4.10%) high mild
  67 (6.70%) high severe
db_put_non_durable/put/key_16B_value_256B
                        time:   [6.7484 µs 6.7684 µs 6.7893 µs]
                        thrpt:  [147.29 Kelem/s 147.75 Kelem/s 148.18 Kelem/s]
Found 90 outliers among 1000 measurements (9.00%)
  6 (0.60%) low mild
  47 (4.70%) high mild
  37 (3.70%) high severe
db_put_non_durable/put/key_16B_value_1024B
                        time:   [6.9436 µs 7.0581 µs 7.1855 µs]
                        thrpt:  [139.17 Kelem/s 141.68 Kelem/s 144.02 Kelem/s]
Found 131 outliers among 1000 measurements (13.10%)
  6 (0.60%) low severe
  11 (1.10%) low mild
  56 (5.60%) high mild
  58 (5.80%) high severe
```
I think that makes sense given this is effectively just testing memtable inserts.

Regarding whether to run in per-commit, I am wondering how meaningful having the benchmark actually is given the potential instability in PRs it'll introduce (and potential burden of keeping an insanely high bar on this specific microbench) -- as noted in Discord, it seems like a regression of ~30x in this benchmark ends up being barely perceptible in the bencher which is a much more realistic workload. 

I think it's worth adding it to CI/CD for now and seeing how it affects our ability to iterate and what rate of false positives are.

**Comment by flaneur2020** [comment]
i think this benchmark can also be regarded a soft limit, like we can output the performance difference compared to the main branch in the PR comments. before we figure out a stable hard threshold, our eyes in code review can also help us spot dramatic performance regressions.

**Review by flaneur2020 — APPROVED** [review]

**Comment by criccomini** [comment]
Oh, that's a great idea! Lemme see if there's a GH action for that. Otherwise, I can build one.

**Comment by criccomini** [comment]
This looks promising:

https://github.com/benchmark-action/github-action-benchmark

File: benches/db_operations.rs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+            let value = b"value";
+            db.put_with_options(
+                key,
+                value,
+                &PutOptions::default(),
+                &WriteOptions {
+                    await_durable: false,
+                },
+            )
+            .await
```
Comment (rodesai) [inline benches/db_operations.rs:27]: should we call `expect` on this to make sure we are not hitting the error path?

**Review by rodesai — COMMENTED** [review]
Looks good! I'm +1 for not setting a hard-bound on how much we allow this regress for a PR. As @agavra said, it's not clear that a regression on this benchmark will cause a meaningful performance degradation in a real use-case. Lets start with eyeballing it.

**Review by dmvk — APPROVED** [review]
Nice!

+1 on adding https://github.com/benchmark-action/github-action-benchmark, it seems to be a promising sanity check

**Comment by criccomini** [comment]
Fixing the clippy issues here: https://github.com/slatedb/slatedb/pull/377

**Pull request merged**